tools/x-repo-types/xrt_tmpl.go

```diff
@@ -0,0 +1,525 @@
+
+// Targets returns a slice of Targets, one for each child of the Type.
+// In the case of structs, the order is the same as the order of the fields in the struct.
+// In the case of maps, the key Target precedes the value Target.
+func (t *Type) Targets() []Target {
+	targets := make([]Target, 0, len(t.Edges))
+	for _, edge := range t.Edges {
+		targets = append(targets, Target{edge, (*t.children)[edge.String()]})
+	}
+	return targets
+}
+
+// IsLeaf returns true if the Type has no children.
+func (t *Type) IsLeaf() bool {
+	return t.children == nil
+}
+
+// Build the Type tree by finding all child types that belong to a kind and recursively
+// building their Type trees.
+func (t *Type) Build() {
+	switch t.Kind {
+	case reflect.Struct:
+		t.buildStructChildren()
+	case reflect.Slice, reflect.Array:
+		t.buildListChild()
+	case reflect.Map:
+		t.buildMapChildren()
+	case reflect.Ptr:
+		t.buildPtrChild()
+	}
+}
+
+func (t *Type) appendChild(typeName, typeTag string, child Type) {
+	edge := Edge{typeName, typeTag}
+	t.Edges = append(t.Edges, edge)
+	if t.children == nil {
+		children := make(Children)
+		t.children = &children
+	}
+	(*t.children)[edge.String()] = child
+}
+
+func (t *Type) buildStructChildren() {
+	for i := 0; i < t.Type.NumField(); i++ {
+		typeField := t.Type.Field(i)
+		typeName := typeField.Name
+
+		// probably we need to skip typeField.Tag == `codec:"-"` as well
+		if typeName == "" || (!unicode.IsUpper(rune(typeName[0])) && typeName != "_struct") {
+			continue
+		}
+
+		typeTag := string(typeField.Tag)
+		child := Type{t.Depth + 1, typeField.Type, typeField.Type.Kind(), nil, nil}
+		child.Build()
+		t.appendChild(typeName, typeTag, child)
+	}
+}
+
+func (t *Type) buildListChild() {
+	tt := t.Type.Elem()
+	child := Type{t.Depth + 1, tt, tt.Kind(), nil, nil}
+	child.Build()
+	t.appendChild("<list elt>", "", child)
+}
+
+func (t *Type) buildMapChildren() {
+	keyType, valueType := t.Type.Key(), t.Type.Elem()
+
+	keyChild := Type{t.Depth + 1, keyType, keyType.Kind(), nil, nil}
+	keyChild.Build()
+	t.appendChild("<map key>", "", keyChild)
+
+	valChild := Type{t.Depth + 1, valueType, valueType.Kind(), nil, nil}
+	valChild.Build()
+	t.appendChild("<map elt>", "", valChild)
+}
+
+func (t *Type) buildPtrChild() {
+	tt := t.Type.Elem()
+	child := Type{t.Depth + 1, tt, tt.Kind(), nil, nil}
+	child.Build()
+	t.appendChild("<ptr elt>", "", child)
+}
+
+// Visit traverses the Target tree and applies any actions provided at each node.
+func (tgt *Target) Visit(actions ...func(Target)) {
+	if len(actions) > 0 {
+		for _, action := range actions {
+			action(*tgt)
+		}
+		for _, target := range tgt.Type.Targets() {
+			target.Visit(actions...)
+		}
+	}
+}
+
+// SerializationDiff recursively computes a diff between two Target's x and y considering only data
+// that impacts serialization, but ignoring any assumption on key ordering.
+func SerializationDiff(x, y Target, exclusions map[string]bool) (*Diff, error) {
```

Comment:
please add a few test cases for this.

Reply:
332dd739dc6bc31caf69a701afe4a3dfebabec2f